File: services/shared/observability/python/metrics.py
"""
Prometheus metrics for ResumeLM Python microservices.

Exposes a /metrics endpoint using the prometheus_client library.

Usage:

    # In main.py:
    from services.shared.observability.python.metrics import (
        add_metrics_route,
        request_count,
        request_latency,
    )

    add_metrics_route(app)

    # In a route or middleware:
    with request_latency.labels(service="auth-service", method="GET", route="/health").time():
        ...
    request_count.labels(
        service="auth-service", method="GET", route="/health", status_code="200"
    ).inc()
"""

from __future__ import annotations

import os
import time
from typing import Callable

from fastapi import FastAPI, Request, Response


def _prometheus_available() -> bool:
    try:
        import prometheus_client  # noqa: F401

        return True
    except ImportError:
        return False


def add_metrics_middleware(app: FastAPI, service_name: str) -> None:
    """
    Register a Starlette middleware that records per-request Prometheus metrics.
    If prometheus_client is not installed the function is a no-op.
    """
    if not _prometheus_available():
        return

    from prometheus_client import Counter, Histogram

    REQUEST_COUNT = Counter(
        "http_requests_total",
        "Total HTTP requests",
        ["service", "method", "route", "status_code"],
    )
    REQUEST_LATENCY = Histogram(
        "http_request_duration_seconds",
        "HTTP request latency",
        ["service", "method", "route"],
        buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10),
    )

    @app.middleware("http")
    async def _metrics_middleware(
        request: Request, call_next: Callable[[Request], Response]
    ) -> Response:
        start = time.perf_counter()
        response: Response = await call_next(request)
        duration = time.perf_counter() - start

        route = request.url.path
        REQUEST_LATENCY.labels(
            service=service_name, method=request.method, route=route
        ).observe(duration)
        REQUEST_COUNT.labels(
            service=service_name,
            method=request.method,
            route=route,
            status_code=str(response.status_code),
        ).inc()

        return response


def add_metrics_route(app: FastAPI) -> None:
    """
    Add a GET /metrics endpoint that serves Prometheus text-format metrics.
    If prometheus_client is not installed, /metrics returns a 501.
    """
    if _prometheus_available():
        from prometheus_client import CONTENT_TYPE_LATEST, generate_latest

        @app.get("/metrics", include_in_schema=False)
        async def metrics() -> Response:
            return Response(
                content=generate_latest(),
                media_type=CONTENT_TYPE_LATEST,
            )

    else:

        @app.get("/metrics", include_in_schema=False)
        async def metrics_unavailable() -> Response:  # type: ignore[misc]
            return Response(
                content="prometheus_client not installed",
                status_code=501,
                media_type="text/plain",
            )

File: services/shared/observability/metrics.ts
/**
 * @module @resume-lm/shared/observability/metrics
 *
 * Lightweight Prometheus metrics helpers for Node.js / Express microservices.
 *
 * Usage:
 * ```ts
 * import { createMetricsMiddleware, metricsHandler } from
 *   '@resume-lm/shared/observability/metrics';
 *
 * // In app setup:
 * app.use(createMetricsMiddleware('resume-service'));
 * app.get('/metrics', metricsHandler);
 * ```
 *
 * Exposes the following metrics on `GET /metrics`:
 *   http_requests_total{service, method, route, status_code}  – counter
 *   http_request_duration_seconds{service, method, route}     – histogram
 *   nodejs_process_info{service, version, node_version}       – gauge (1)
 */

import type { NextFunction, Request, Response } from 'express';

// ---------------------------------------------------------------------------
// Minimal Prometheus text-format helpers (no external dependency required)
// ---------------------------------------------------------------------------

interface Counter {
  labels: Record<string, string>;
  value: number;
}

interface HistogramBucket {
  le: number | '+Inf';
  count: number;
}

interface HistogramSample {
  labels: Record<string, string>;
  sum: number;
  count: number;
  buckets: HistogramBucket[];
}

const requestCounters: Counter[] = [];
const requestHistograms: HistogramSample[] = [];

const DURATION_BUCKETS = [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10];

function labelKey(labels: Record<string, string>): string {
  return Object.entries(labels)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `${k}="${v}"`)
    .join(',');
}

function incrementCounter(labels: Record<string, string>): void {
  const key = labelKey(labels);
  const existing = requestCounters.find((c) => labelKey(c.labels) === key);
  if (existing) {
    existing.value += 1;
  } else {
    requestCounters.push({ labels, value: 1 });
  }
}

function observeHistogram(labels: Record<string, string>, value: number): void {
  const key = labelKey(labels);
  let sample = requestHistograms.find((h) => labelKey(h.labels) === key);
  if (!sample) {
    sample = {
      labels,
      sum: 0,
      count: 0,
      buckets: (DURATION_BUCKETS.map((le) => ({ le, count: 0 })) as HistogramBucket[]).concat([
        { le: '+Inf', count: 0 },
      ]),
    };
    requestHistograms.push(sample);
  }
  sample.sum += value;
  sample.count += 1;
  for (const bucket of sample.buckets) {
    if (bucket.le === '+Inf' || value <= bucket.le) {
      bucket.count += 1;
    }
  }
}

function formatLabels(labels: Record<string, string>): string {
  const parts = Object.entries(labels).map(([k, v]) => `${k}="${v}"`);
  return parts.length > 0 ? `{${parts.join(',')}}` : '';
}

function renderMetrics(serviceName: string): string {
  const lines: string[] = [];

  // ── http_requests_total ────────────────────────────────────────────────────
  lines.push('# HELP http_requests_total Total HTTP requests processed');
  lines.push('# TYPE http_requests_total counter');
  for (const c of requestCounters) {
    lines.push(`http_requests_total${formatLabels(c.labels)} ${c.value}`);
  }

  // ── http_request_duration_seconds ─────────────────────────────────────────
  lines.push('# HELP http_request_duration_seconds HTTP request latency in seconds');
  lines.push('# TYPE http_request_duration_seconds histogram');
  for (const h of requestHistograms) {
    const baseLabels = formatLabels(h.labels);
    for (const bucket of h.buckets) {
      const le = bucket.le === '+Inf' ? '+Inf' : String(bucket.le);
      const labelStr = baseLabels
        ? baseLabels.slice(0, -1) + `,le="${le}"}`
        : `{le="${le}"}`;
      lines.push(`http_request_duration_seconds_bucket${labelStr} ${bucket.count}`);
    }
    lines.push(`http_request_duration_seconds_sum${baseLabels} ${h.sum}`);
    lines.push(`http_request_duration_seconds_count${baseLabels} ${h.count}`);
  }

  // ── nodejs_process_info ────────────────────────────────────────────────────
  lines.push('# HELP nodejs_process_info Node.js process information');
  lines.push('# TYPE nodejs_process_info gauge');
  lines.push(
    `nodejs_process_info{service="${serviceName}",` +
      `node_version="${process.version}"} 1`,
  );

  // ── process_uptime_seconds ────────────────────────────────────────────────
  lines.push('# HELP process_uptime_seconds Process uptime in seconds');
  lines.push('# TYPE process_uptime_seconds gauge');
  lines.push(`process_uptime_seconds{service="${serviceName}"} ${process.uptime()}`);

  return lines.join('\n') + '\n';
}

// ---------------------------------------------------------------------------
// Middleware + handler
// ---------------------------------------------------------------------------

/**
 * Express middleware that records HTTP request metrics.
 * Mount this before your routes.
 */
export function createMetricsMiddleware(
  serviceName: string,
): (req: Request, res: Response, next: NextFunction) => void {
  return (req: Request, res: Response, next: NextFunction): void => {
    const start = process.hrtime.bigint();

    res.on('finish', () => {
      const durationNs = process.hrtime.bigint() - start;
      const durationS = Number(durationNs) / 1e9;

      // Normalise route: replace UUIDs and numeric IDs with placeholders
      const route = req.route?.path ?? req.path.replace(/\/[0-9a-f-]{8,}/gi, '/:id');

      const labels: Record<string, string> = {
        service: serviceName,
        method: req.method,
        route,
        status_code: String(res.statusCode),
      };

      incrementCounter(labels);
      observeHistogram(
        { service: serviceName, method: req.method, route },
        durationS,
      );
    });

    next();
  };
}

/**
 * Express route handler that serves the Prometheus text format on `GET /metrics`.
 */
export function metricsHandler(
  _req: Request,
  res: Response,
  _next: NextFunction,
  serviceName: string,
): void {
  res.set('Content-Type', 'text/plain; version=0.0.4; charset=utf-8');
  res.end(renderMetrics(serviceName));
}
